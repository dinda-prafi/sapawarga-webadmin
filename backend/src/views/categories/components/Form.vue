<template>
  <div>
    <el-row :gutter="20">
      <el-col :xs="24" :sm="16">
        <el-form ref="form" :model="form" :rules="rules" :status-icon="true" label-width="160px">

          <el-form-item label="Nama Kategori" :prop="validateName">
            <el-input v-model="form.name" type="text" placeholder="Nama Kategori" @focus="changePropName" />
          </el-form-item>

          <el-form-item label="Fitur" prop="type">
            <el-select v-model="form.type" placeholder="Pilih Fitur">
              <el-option
                v-for="item in optionType"
                :key="item.id"
                :label="item.name"
                :value="item.id"
              />
            </el-select>
          </el-form-item>

          <el-form-item>
            <el-button type="primary" :loading="loading" @click="submitProcess">
              <span v-if="isEdit === true">{{ $t('crud.save-update') }}</span>
              <span v-else>{{ $t('crud.save-publish') }}</span>
            </el-button>
            <router-link :to="'/categories/index'">
              <el-button type="info">{{ $t('crud.cancel') }}</el-button>
            </router-link>
          </el-form-item>

        </el-form>
      </el-col>
    </el-row>
  </div>
</template>

<script>
import { fetchRecord, create, update, fetchList } from '@/api/categories'
import { containsWhitespace } from '@/utils/validate'
import checkPermission from '@/utils/permission'

const defaultForm = {
  name: null,
  type: null,
  status: 10
}

export default {
  props: {
    isEdit: {
      type: Boolean,
      default: false
    }
  },
  data() {
    const validatorTitleWhitespace = (rule, value, callback) => {
      if (containsWhitespace(value) === true) {
        callback(new Error('Nama kategori yang diisi tidak valid.'))
      }

      callback()
    }

    return {
      form: Object.assign({}, defaultForm),
      loading: false,
      validateName: 'name',
      rules: {
        name: [
          { required: true, message: 'Nama kategori harus diisi.', trigger: 'blur' },
          { min: 4, message: 'Nama kategori minimal 4 karakter', trigger: 'blur' },
          { max: 25, message: 'Nama kategori maksimal 25 karakter', trigger: 'blur' },
          { validator: validatorTitleWhitespace, trigger: 'blur' }
        ],
        type: [
          { required: true, message: 'Fitur harus diisi.', trigger: 'change' }
        ],
        errorName: [
          { required: true, message: 'Nama kategori sudah digunakan', trigger: 'change' }
        ]
      },
      listQuery: {
        type: null
      },
      tempRoute: {},
      optionType: []
    }
  },
  mounted() {
    if (this.isEdit) {
      const id = this.$route.params && this.$route.params.id
      this.fetchData(id)
    } else {
      this.form = Object.assign({}, defaultForm)
    }
    this.fetchDataTypes()
  },
  methods: {
    fetchData(id) {
      fetchRecord(id).then(response => {
        this.form = response.data
      }).catch(err => console.log(err))
    },

    fetchDataTypes() {
      if (checkPermission(['staffSaberhoax'])) {
        this.listQuery['type'] = 'newsHoax'
      }
      fetchList(this.listQuery).then(response => {
        const { data } = response
        this.optionType = data.items.map(item => {
          return {
            id: item.id,
            name: item.name
          }
        })
      }).catch(err => console.error(err))
    },

    async submitProcess() {
      const valid = await this.$refs.form.validate()

      if (!valid) {
        return
      }

      try {
        this.loading = true

        const data = {}

        Object.assign(data, this.form)

        if (this.isEdit) {
          const id = this.$route.params && this.$route.params.id

          await update(id, data)

          this.$message.success(this.$t('crud.update-success'))

          this.$router.push('/categories/index')
        } else {
          await create(data)
          this.$message.success(this.$t('crud.create-success'))

          this.$router.push('/categories/index')
        }
      } catch (error) {
        const messageApi = error.response.data.data
        const messageList = []
        let messageJoin = null

        messageApi.forEach(element => {
          messageList.push(element.message)
          messageJoin = messageList.join(' Dan ')
        })
        this.$message.error(this.$t(messageJoin))
        if (messageApi[0].field === 'name') {
          this.form.name = null
          this.validateName = 'errorName'
        }
      } finally {
        this.loading = false
      }
    },
    changePropName() {
      if (this.validateName === 'errorName') {
        this.validateName = 'name'
      } else {
        this.validateName = 'name'
      }
    }
  }
}
</script>
