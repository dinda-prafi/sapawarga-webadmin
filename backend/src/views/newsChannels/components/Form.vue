<template>
  <div>
    <el-row :gutter="20">
      <el-col :xs="24" :sm="16">
        <el-form ref="form" :model="form" :rules="rules" :status-icon="true" label-width="210px">

          <el-form-item label="Sumber Berita" prop="name">
            <el-input v-model="form.name" type="text" placeholder="Sumber Berita" />
          </el-form-item>

          <el-form-item label="Tautan Logo Sumber Berita" prop="icon_url">
            <el-input v-model="form.icon_url" type="text" placeholder="http://www" />
          </el-form-item>

          <el-form-item label="Tautan" prop="website">
            <el-input v-model="form.website" type="text" placeholder="http://www" />
          </el-form-item>

          <el-form-item>
            <el-button type="primary" :loading="loading" @click="submitProcess">{{ $t('crud.save-publish') }}</el-button>

            <router-link :to="'/news-channels/index'">
              <el-button type="info">{{ $t('crud.cancel') }}</el-button>
            </router-link>
          </el-form-item>

        </el-form>
      </el-col>
    </el-row>
  </div>
</template>

<script>
import { fetchRecord, create } from '@/api/newsChannels'
import { containsWhitespace, validUrl } from '@/utils/validate'

const defaultForm = {
  name: null,
  website: null,
  icon_url: null,
  status: 10
}

export default {
  props: {
    isEdit: {
      type: Boolean,
      default: false
    }
  },
  data() {
    const validatorTextWhitespace = (rule, value, callback) => {
      if (containsWhitespace(value) === true) {
        callback(new Error('Teks yang diisi tidak valid.'))
      }

      callback()
    }

    const validatorUrl = (rule, value, callback) => {
      if (validUrl(value) === false) {
        callback(new Error('URL tidak valid.'))
      }

      callback()
    }

    return {
      form: Object.assign({}, defaultForm),
      loading: false,
      rules: {
        name: [
          { required: true, message: 'Judul sumber berita harus diisi.', trigger: 'blur' },
          { min: 5, message: 'Judul sumber berita minimal 5 karakter', trigger: 'blur' },
          { max: 100, message: 'Judul sumber berita maksimal 100 karakter', trigger: 'blur' },
          { validator: validatorTextWhitespace, trigger: 'blur' }
        ],
        website: [
          { required: true, message: 'URL harus diisi.', trigger: 'blur' },
          { validator: validatorUrl, trigger: 'blur' }
        ],
        icon_url: [
          { required: true, message: 'URL harus diisi.', trigger: 'blur' },
          { validator: validatorUrl, trigger: 'blur' }
        ]
      }
    }
  },
  mounted() {
    if (this.isEdit) {
      const id = this.$route.params && this.$route.params.id
      this.fetchData(id)
    } else {
      this.form = Object.assign({}, defaultForm)
    }
  },
  methods: {
    fetchData(id) {
      fetchRecord(id).then(response => {
        this.form = response.data
      }).catch(err => console.log(err))
    },

    async submitProcess() {
      const valid = await this.$refs.form.validate()

      if (!valid) {
        return
      }

      try {
        this.loading = true

        const data = {}

        Object.assign(data, this.form)

        if (this.isEdit) {
          // TODO: edit
          /* const id = this.$route.params && this.$route.params.id

          await update(id, data)

          this.$message.success(this.$t('crud.update-success'))

          this.$router.push('/news-channels/index') */
        } else {
          await create(data)
          this.$message.success(this.$t('crud.create-success'))

          this.$router.push('/news-channels/index')
        }
      } catch (error) {
        const errorLength = error.response.data.data.length
        const name = error.response.data.data[0].field
        const message = error.response.data.data[0].message
        if (errorLength === 2) {
          this.$message.error(this.$t('Sumber Berita dan Tautan sudah digunakan'))
        } else {
          this.$message.error(this.$t(message))
        }
      } finally {
        this.loading = false
      }
    }
  }
}
</script>
