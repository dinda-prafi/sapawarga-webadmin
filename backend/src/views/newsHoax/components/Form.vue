<template>
  <div class="components-container">
    <el-row :gutter="20">
      <el-col :xs="24" :sm="8" :lg="5">
        <AttachmentPhotoUpload type="news_photo" :initial-url="news.cover_path_url" style="margin-bottom: 25px" name="news_photo" @onUpload="photoUploaded" />
      </el-col>
      <el-col :xs="24" :sm="16" :lg="19">

        <el-form ref="news" :model="news" :rules="rules" :status-icon="true" label-width="160px">
          <el-form-item label="Judul Berita" prop="title">
            <el-input v-model="news.title" name="title" type="text" placeholder="Judul Berita" />
          </el-form-item>

          <el-form-item label="Kategori" prop="category_id">
            <el-select v-model="news.category" name="category" placeholder="Pilih Kategori">
              <el-option value="1" label="Disinformasi" />
              <el-option value="2" label="Fakta" />
              <el-option value="3" label="Klarifikasi" />
            </el-select>
          </el-form-item>

          <el-form-item label="Konten Berita" prop="content">
            <div>
              <tinymce v-model="news.content" :height="300" name="content" />
            </div>
          </el-form-item>

          <el-form-item>
            <el-button v-if="isEdit" type="primary" :loading="loading" @click="submitForm">{{ $t('crud.save-update') }}</el-button>
            <el-button v-else type="primary" :loading="loading" @click="submitForm">{{ $t('crud.save-news') }}</el-button>

            <router-link :to="'/news/index'">
              <el-button type="info">{{ $t('crud.cancel') }}</el-button>
            </router-link>
          </el-form-item>

        </el-form>

      </el-col>
    </el-row>
  </div>
</template>

<script>
import AttachmentPhotoUpload from '@/components/AttachmentPhotoUpload'
import { containsWhitespace } from '@/utils/validate'
import { fetchRecord, create, update } from '@/api/newsHoax'
import Tinymce from '@/components/Tinymce'
import moment from 'moment'
import { mapGetters } from 'vuex'

export default {
  components: { Tinymce, AttachmentPhotoUpload },
  props: {
    isEdit: {
      type: Boolean,
      default: false
    }
  },
  data() {
    const validatorTitleWhitespace = (rule, value, callback) => {
      if (containsWhitespace(value) === true) {
        callback(new Error('Judul Berita yang diisi tidak valid'))
      }
      callback()
    }

    const validatorTitleWhitespaceContent = (rule, value, callback) => {
      if (containsWhitespace(value) === true) {
        callback(new Error('Konten Berita yang diisi tidak valid'))
      }
      callback()
    }

    return {
      loading: false,
      news: {
        title: null,
        category_id: null,
        category: null,
        source_date: moment().startOf('day'),
        content: null,
        seq: null,
        status: null
      },
      area: null,
      rules: {
        title: [
          {
            required: true,
            message: 'Judul Berita harus diisi',
            trigger: 'blur'
          },
          {
            min: 10,
            message: 'Judul Berita minimal 10 karakter',
            trigger: 'blur'
          },
          {
            max: 100,
            message: 'Judul Berita maksimal 100 karakter',
            trigger: 'blur'
          },
          {
            validator: validatorTitleWhitespace,
            trigger: 'blur'
          }
        ],
        category_id: [
          {
            required: true,
            message: 'Sumber Berita harus diisi',
            trigger: 'change'
          }
        ],
        content: [
          {
            required: true,
            message: 'Konten Berita harus diisi',
            trigger: 'blur'
          },
          {
            validator: validatorTitleWhitespaceContent,
            trigger: 'blur'
          }
        ]
      }
    }
  },
  computed: {
    ...mapGetters([
      'user_id'
    ])
  },
  created() {
    if (this.isEdit) {
      const id = this.$route.params && this.$route.params.id
      this.fetchData(id)
    }
  },
  methods: {
    fetchData(id) {
      fetchRecord(id).then(response => {
        this.news = response.data
      }).catch(err => {
        console.log(err)
      })
    },

    photoUploaded(path, url) {
      this.news.cover_path = path
    },

    async submitForm() {
      const valid = await this.$refs.news.validate()

      if (!valid) {
        return
      }

      try {
        this.loading = true

        const data = {}

        Object.assign(data, this.news)
        data.source_date = moment(data.source_date).format('YYYY-MM-DD')

        if (this.isEdit) {
          const id = this.$route.params && this.$route.params.id
          await update(id, data)

          this.$message.success(this.$t('crud.update-success'))

          this.$router.push('/news-hoax/index')
        } else {
          data.status = 10
          await create(data)

          this.$message.success(this.$t('crud.create-success'))

          this.$router.push('/news-hoax/index')
        }
      } catch (e) {
        const imageError = e.response.data.data[0].field
        if (imageError === 'cover_path') {
          this.$message.error(this.$t('errors.news-image-null'))
        }
      } finally {
        this.loading = false
      }
    }
  }
}
</script>
