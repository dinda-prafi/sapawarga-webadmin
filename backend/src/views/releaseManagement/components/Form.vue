<template>
  <div name="release-management-form">
    <el-row :gutter="20">
      <el-col :xs="24" :sm="16">
        <el-form ref="form" :model="form" :rules="rules" :status-icon="true" label-width="210px">

          <el-form-item label="Version" :prop="validateVersion">
            <el-input v-model="form.version" type="text" placeholder="Version" @focus="changePropVersion" />
          </el-form-item>

          <el-form-item label="Force Update" prop="force_update">
            <el-radio-group v-model="form.force_update">
              <el-radio :label="true">True</el-radio>
              <el-radio :label="false">False</el-radio>
            </el-radio-group>
          </el-form-item>

          <el-form-item>
            <el-button type="primary" :loading="loading" @click="submitProcess">
              <span v-if="isEdit === true">{{ $t('crud.save-update') }}</span>
              <span v-else>{{ $t('crud.save-create') }}</span>
            </el-button>
            <router-link :to="'/release-management/index'">
              <el-button type="info">{{ $t('crud.cancel') }}</el-button>
            </router-link>
          </el-form-item>

        </el-form>
      </el-col>
    </el-row>
  </div>
</template>

<script>
import { fetchRecord, update, create } from '@/api/releaseManagement'
import { containsWhitespace } from '@/utils/validate'

const defaultForm = {
  version: null,
  force_update: null
}

export default {
  props: {
    isEdit: {
      type: Boolean,
      default: false
    }
  },
  data() {
    const validatorTextWhitespace = (rule, value, callback) => {
      if (containsWhitespace(value) === true) {
        callback(new Error('Teks yang diisi tidak valid.'))
      }

      callback()
    }

    return {
      form: Object.assign({}, defaultForm),
      loading: false,
      validateVersion: 'version',
      rules: {
        version: [
          { required: true, message: 'Version harus diisi.', trigger: 'blur' },
          { validator: validatorTextWhitespace, trigger: 'blur' }
        ],
        errorVersion: [
          { required: true, message: 'Version sudah digunakan', trigger: 'change' }
        ]
      }
    }
  },
  mounted() {
    if (this.isEdit) {
      const id = this.$route.params && this.$route.params.id
      this.fetchData(id)
    } else {
      this.form = Object.assign({}, defaultForm)
    }
  },
  methods: {
    fetchData(id) {
      fetchRecord(id).then(response => {
        this.form = response.data
      }).catch(err => console.log(err))
    },

    async submitProcess() {
      const valid = await this.$refs.form.validate()

      if (!valid) {
        return
      }

      try {
        this.loading = true

        const data = {}

        Object.assign(data, this.form)

        if (this.isEdit) {
          // TODO: edit
          const id = this.$route.params && this.$route.params.id

          await update(id, data)

          this.$message.success(this.$t('crud.update-success'))
          this.$router.push('/release-management/index')
        } else {
          await create(data)
          this.$message.success(this.$t('crud.create-success'))

          this.$router.push('/release-management/index')
        }
      } catch (error) {
        const messageApi = error.response.data.data
        const messageList = []
        let messageJoin = null

        messageApi.forEach(element => {
          messageList.push(element.message)
          messageJoin = messageList.join(' Dan ')
        })
        this.$message.error(this.$t(messageJoin))

        if (messageApi[0].field === 'version') {
          this.form.version = null
          this.validateVersion = 'errorVersion'
        }
      } finally {
        this.loading = false
      }
    },
    changePropVersion() {
      if (this.validateVersion === 'errorVersion') {
        this.validateVersion = 'version'
      } else {
        this.validateVersion = 'version'
      }
    }

  }
}
</script>
