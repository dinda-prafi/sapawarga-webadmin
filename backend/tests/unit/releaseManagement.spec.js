import { shallowMount, createLocalVue } from '@vue/test-utils'
import ElementUI from 'element-ui';
import releaseManagementCreate from '@/views/releaseManagement/create'
import releaseManagementEdit from '@/views/releaseManagement/edit'
import releaseManagementList from '@/views/releaseManagement/list'
import Form from '@/views/releaseManagement/components/Form'

const localVue = createLocalVue()
localVue.use(ElementUI)

beforeEach(() => {
  // Clear all instances and calls to constructor and all methods:
  jest.resetModules()
  jest.clearAllMocks()
})

describe('Categories', () => {
  const build = () => {
    // arrange
    const wrapper = shallowMount(releaseManagementList, { localVue })
    return {
      wrapper
    }
  }

  it('call create form', () => {
    const wrapper = shallowMount(releaseManagementCreate, {
      localVue
    })

    expect(wrapper.find(Form).exists()).toBe(true)
  })

  it('call edit form', () => {
    const wrapper = shallowMount(releaseManagementEdit, {
      localVue
    })

    expect(wrapper.find(Form).exists()).toBe(true)
  })

  it('renders the component', () => {
    const { wrapper } = build()
    // assert
    // expect(wrapper.html()).toMatchSnapshot()
  })
})
