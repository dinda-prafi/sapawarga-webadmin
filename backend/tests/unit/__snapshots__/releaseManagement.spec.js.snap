// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories renders the component 1`] = `
<div name="release-management-list" class="app-container">
  <el-row-stub tag="div" gutter="20" justify="start" align="top">
    <el-col-stub span="24" tag="div" lg="24">
      <el-row-stub tag="div" justify="start" align="top" style="margin: 10px 0px;">
        <el-col-stub span="12" tag="div">
          <router-link-stub to="[object Object]" tag="a" event="click">
            <el-button-stub type="primary" size="small" icon="el-icon-plus" nativetype="button">
              Tambah Release
            </el-button-stub>
          </router-link-stub>
        </el-col-stub>
      </el-row-stub>
      <el-table-stub fit="true" stripe="true" border="true" showheader="true" highlightcurrentrow="true" selectonindeterminate="true" indent="16" style="width: 100%;">
        <el-table-column-stub type="index" width="50" resizable="true" align="center" filtermultiple="true" index="function () { [native code] }" sortorders="ascending,descending,"></el-table-column-stub>
        <el-table-column-stub type="default" label="Version" prop="version" minwidth="150" sortable="custom" resizable="true" filtermultiple="true" sortorders="ascending,descending,"></el-table-column-stub>
        <el-table-column-stub type="default" label="Force Update" prop="force_update" minwidth="150" sortable="custom" resizable="true" formatter="function () { [native code] }" filtermultiple="true" sortorders="ascending,descending,"></el-table-column-stub>
        <el-table-column-stub type="default" label="Actions" minwidth="150px" resizable="true" align="center" filtermultiple="true" sortorders="ascending,descending,"></el-table-column-stub>
      </el-table-stub>
      <pagination-stub total="0" page="1" limit="10" pagesizes="10,20,30,50" layout="total, sizes, prev, pager, next, jumper" background="true" autoscroll="true" style="display: none;"></pagination-stub>
    </el-col-stub>
  </el-row-stub>
</div>
`;
